### DISCUSSION TITLE
How to calculate the mapping between time derivatives of generalized position and generalized velocity?

### DISCUSSION THREAD

**Discussion by Zionshang** [stack-of-tasks/pinocchio#2493] [Q&A] [answered]
The generalized position and generalized velocity are $\boldsymbol{q}\in \mathbb{R} ^{n_q}$ and $\boldsymbol{v}\in \mathbb{R} ^{n_v}$, respectively. The relationship of time derivatives of generalized position and generalized velocity is 

$\dot{\boldsymbol{q}}=N\left( \boldsymbol{q} \right) \boldsymbol{v}$.

$N\left( \boldsymbol{q} \right) \in \mathbb{R} ^{n_q\times n_v}$ is the identity matrix in most cases, but not for quaternion Degrees of Freedom.

So my question is: how to calculate $N\left( \boldsymbol{q} \right)$.

Although I know that $N(q)$ can be calculated by the relationship between time derivatives of quaternion and the angular velocity, I wonder if there is a uniform formula for calculating N(q).

Computational $N(q)$ is important for trajectory optimization, because if the generalized position on the discrete trajectory is known, then the generalized velocity can be calculated by $N(q)$ .

I would appreciate it if you could answer my question!

**Comment by ManifoldFR** [comment]
Hello, thanks for your interest in Pinocchio!

There is no matrix $N(q)$ in Pinocchio, because $q$ is directly an element of the appropriate configuration (matrix) Lie group, and generalized velocities $v$ are expressed in the Lie algebra. What that changes, however, is that the ODE you want to express is a Lie group ODE $\dot{q} = q\cdot v$ where $\omega\mapsto q\cdot\omega$ is the left group action.

**Reply by ManifoldFR** [reply, marked as answer]
Any numerical solver which wants to use non-Euclidean Lie group configurations and Pinocchio has to perform the necessary steps (trivializing gradients, using the exp map given by `pin.integrate(model, q, dq)` $=q\cdot \exp_Q(\delta q)$ to perform gradient descent, etc).

Examples of this use are our NLP solver [proxsuite-nlp](https://github.com/Simple-Robotics/proxsuite-nlp/), and the trajopt libraries [crocoddyl](https://github.com/loco-3d/crocoddyl/) and [aligator](https://github.com/Simple-Robotics/aligator/).

**Reply by Zionshang** [reply]
Thanks for your help!!!!!

**Reply by ManifoldFR** [reply]
没有问题！Good luck 😃
